packages/ovee/src/core/TemplateComponent.ts

```diff
@@ -1,54 +1,90 @@
 import { html, render, TemplateResult } from 'lit-html';
-
-import watch from 'src/decorators/watch';
+import { doWatchEffect, ref } from 'src/reactive';
+import { Task } from 'src/utils';
 
 import App from './App';
 import Component, { ComponentOptions } from './Component';
 import * as protectedFields from './protectedFields';
 
 export default class TemplateComponent extends Component {
-    [protectedFields.UPDATE_AF]: number | null = null;
-    [protectedFields.UPDATE_PROMISE]: Promise<void> | null = null;
-
-    readonly html!: typeof html;
-
-    constructor(element: Element, app: App, options?: ComponentOptions) {
-        super(element, app, options);
-
-        Object.defineProperty(this, 'html', {
-            value: html,
-            writable: false,
-            configurable: false
-        });
-    }
-
-    async [protectedFields.BEFORE_INIT](): Promise<void> {
-        await super[protectedFields.BEFORE_INIT]();
-        await this.$requestUpdate();
-    }
-
-    template(): TemplateResult | string {
-        return this.html``;
-    }
-
-    @watch('*')
-    $requestUpdate(): Promise<void> {
-        if (!this[protectedFields.UPDATE_AF]) {
-            this[protectedFields.UPDATE_PROMISE] = new Promise((resolve) => {
-                this[protectedFields.UPDATE_AF] = requestAnimationFrame(() => {
-                    this[protectedFields.UPDATE_AF] = null;
-                    this[protectedFields.RENDER]();
-                    resolve();
-                });
-            });
-        }
-
-        return this[protectedFields.UPDATE_PROMISE] ?? Promise.resolve();
-    }
-
-    [protectedFields.RENDER](): void {
-        render(this.template(), this.$element, {
-            eventContext: this as any
-        });
-    }
+	readonly html!: typeof html;
+
+	[protectedFields.UPDATE_TASK]: Task<void> | null = null;
+
+	private stopWatch = () => {};
+	private updateMarker = ref(false);
+	private compiledTemplate: TemplateResult | string = '';
+
+	constructor(element: Element, app: App, options?: ComponentOptions) {
+		super(element, app, options);
+
+		Object.defineProperty(this, 'html', {
+			value: html,
+			writable: false,
+			configurable: false,
+		});
+	}
+
+	private get updateTask() {
+		return this[protectedFields.UPDATE_TASK];
+	}
+
+	async [protectedFields.BEFORE_INIT](): Promise<void> {
+		await super[protectedFields.BEFORE_INIT]();
+		this.stopWatch = doWatchEffect(
+			() => {
+				this.markRefrashable();
+				this.compiledTemplate = this.template();
+				this.doUpdate();
+			},
+			{ flush: 'sync' }
+		);
+		await this.updateTask;
+	}
+
+	destroy() {
+		this.stopWatch();
+	}
+
+	private markRefrashable() {
+		this.updateMarker.value;
+	}
+
+	private forceRefresh() {
+		this.updateMarker.value = !this.updateMarker.value;
+	}
+
+	private doUpdate(): void {
+		if (this.updateTask) {
+			return;
+		}
+
+		this[protectedFields.UPDATE_TASK] = new Task();
+
+		requestAnimationFrame(() => {
```

Comment:
As the updated reactivity brings much better performance, we might test how per-component `requestAnimationFrame` performs in comparison to a centralized scheduler

Reply:
I created an issue for that: #17 

we should consider if we want to ship it in `v2.1`

Reply:
Let's keep it for future update then